swift/Sources/FlatBuffers/Verifier.swift

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2021 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+
+/// Verifier that check if the buffer passed into it is a valid,
+/// safe, aligned Flatbuffers object since swift read from `unsafeMemory`
+public struct Verifier {
+
+  /// Flag to check for alignment if true
+  fileprivate let _checkAlignment: Bool
+  /// Capacity of the current buffer
+  fileprivate var _capacity: Int
+  /// Current ApparentSize
+  fileprivate var _apparentSize: UOffset = 0
+  /// Amount of tables present within a buffer
+  fileprivate var _tableCount = 0
+
+  /// Capacity of the buffer
+  internal var capacity: Int { _capacity }
+  /// Current reached depth within the buffer
+  internal var _depth = 0
+  /// Current verifiable ByteBuffer
+  internal var _buffer: ByteBuffer
+  /// Options for verification
+  internal let _options: VerifierOptions
+
+  /// Initializer for the verifier
+  /// - Parameters:
+  ///   - buffer: Bytebuffer that is required to be verified
+  ///   - options: `VerifierOptions` that set the rule for some of the verification done
+  ///   - checkAlignment: If alignment check is required to be preformed
+  /// - Throws: `exceedsMaxSizeAllowed` if capacity of the buffer is more than 2GiB
+  public init(
+    buffer: inout ByteBuffer,
```

Comment:
@mustiikhalil , any reason this is an inout parameter? ByteBuffer internally maintains the pointer, so extra copy or not of that pointer should be OK?

Reply:
@liuliu my thought process here is that we don't want to also copy the `ByteBuffer` struct too. Trying to make it as fast as possible. I haven't tried it without `inout`, however if you see that its faster without the inout please create a PR so we can review it together

Reply:
Yeah, it makes some of the wrapping more challenging as I need to pass inout parameters all the way down for ByteBuffer. Do you have a benchmark suite I can run? Definitely can play around with the benchmark and put up a PR for discussion.

Reply:
Before doing so, can you let me know how are you using the verifier? or how does the code that you currently have will function with the current implementation of the verifier?

> Do you have a benchmark suite I can run?

Unfortunately, there isn't.

Reply:
No worries. I can help to build the benchmark.

I started to add persisted dictionary support in Dflat: https://github.com/liuliu/dflat/blob/unstable/src/sqlite/SQLiteWorkspaceDictionary.swift#L107 Because you can put arbitrary objects encoded there, run verify before decode seems to be a good idea.

Yeah, before doing the benchmark, I need to check a few things, seems adding verify breaks my unit tests somehow ...

Reply:
if Dflat isn't going to request data from a network, its safe to assume that the data is always safe.

Reply:
There is a bit more back-story I am thinking through (sorting it out), and seeing `Verifier` support, thinking it is a good idea. If you can bear with me for the long paragraph:

I am currently working on two new things with Dflat, 1. automatically generate flatbuffers schema from GraphQL queries / schema. 2. having a persisted dictionary support that you can save any FlatBuffersCodable object into a dictionary and read them out.

These two independently, are perfectly safe assuming no data corruption on disk.

For 1., even though there is no guarantee on parameter orders when GraphQL schema changes (FlatBuffer's backward compatibility story requires new field to be appended to the end, old field to never be removed, only deprecated), I can generate "digest" that effectively versioning the GraphQL schema into different sqlite backing tables.

For 2, if you have hand-written flatbuffers schemas, you can always guarantee safety if you save the same object with the same key in the persisted dictionary, and update schema based on the rules.

But if you have generated flatbuffers schema from 1, and try to save that into the persisted dictionary in 2., it can be problematic for end-users. A way to avoid this is to use Verifier, another way is to incorporate the "digest" concept into my wrapper data in 2.

As I said, I am still iterating through how to deal with these, and just want to dip my toe with Verifier as I have other non-open-source project actually depends on network transmitted flatbuffers data :)